src/Microsoft.Azure.SignalR.Protocols/ServiceProtocol.cs

```diff
@@ -313,7 +315,15 @@ private static void WriteMultiUserDataMessage(ref MessagePackWriter writer, Mult
         private static void WriteBroadcastDataMessage(ref MessagePackWriter writer, BroadcastDataMessage message)
         {
             writer.WriteArrayHeader(3);
-            writer.Write(ServiceProtocolConstants.BroadcastDataMessageType);
+            if (message.MessageId == null)
+            {
+                writer.Write(ServiceProtocolConstants.BroadcastDataMessageType);
+            }
+            else
+            {
+                writer.Write(ServiceProtocolConstants.BroadcastDataMessageWithMessageIdType);
+                writer.Write(message.MessageId);
+            }

```

Comment:
invalid array header, it should be 4.
we should create a standard for var-length messages.
my prefer: [type, wellknown properties, ["ext-prop1", "value-1", "ext-prop2", "value-2"]]